File: java/sorts/merge_sort.java
import java.util.*;

public class merge_sort{
    public static void main(String[] args){
        int[] arr = new int[]{13,8,9,5,1,4};
        merge_sort m = new merge_sort();
        m.mergeSort(arr);
        for(int i=0;i<arr.length;i++)
            System.out.print(arr[i]+" ");
        System.out.println();
    }

    void mergeSort(int[] arr){
        mergeSortHelper(arr, 0, arr.length);
    }

    void mergeSortHelper(int[] arr, int start, int end){
        if(start >= end-1) return;
        int mid = start + (end - start)/2;
        mergeSortHelper(arr, start, mid);
        mergeSortHelper(arr, mid, end);
        merge(arr, start, mid, end);
    }

    void merge(int[] arr, int start, int mid, int end){
        int[] temp = new int[end-start];
        int i = start;
        int j = mid;
        int k = 0;
        while(i<mid && j<end){
            if(arr[i]<arr[j]){
                temp[k++] = arr[i++];
            }
            else temp[k++] = arr[j++];
        }

        while(i < mid) temp[k++] = arr[i++];
        while(j < end) temp[k++] = arr[j++];
        
        for(int index = 0; index<temp.length; index++) 
            arr[start+index] = temp[index];
    }
}

File: cpp/merge_sort.cpp
#include<iostream>
#include<vector>
using namespace std;

void mergeSort(vector<int>& arr, int start, int end){
    if(end - start <= 1) return;
    int mid = start + (end - start)/2;
    mergeSort(arr, start, mid);
    mergeSort(arr, mid, end);
    //half sorted
    for(int i=mid;i<end;i++) cout<<arr[i]<<" ";
    cout<<endl;
    vector<int> tmp;
    int j = mid;
    for(int i = start; i<mid; i++){
        while(j<end && arr[j] < arr[i]) tmp.push_back(arr[j++]);
        tmp.push_back(arr[i]);
    }
    copy(tmp.begin(),tmp.end(), arr.begin()+start);
}

int main(){
    vector<int> arr = {8,5,6,3,4,2,1,7};
    mergeSort(arr, 0, arr.size());
    for(int i=0;i<arr.size();i++){
        cout<<arr[i]<<" ";
    }
    cout<<endl;
    return 0;
}
